Add the Runtime Cache Design Docs (#764)
luna/enso: docs/runtime/caching.md

    a.quux d
```

> The cache is active for the currently visible scope in Enso Studio, so when a

Author: Can't imagine how I missed that. I feel so absent-minded lately :facepalm:

user enters the function foo, the cache stores the intermediate results in this
function (in this case c and d), as well as the inputs to the function (in this
case a, and b).

All intermediate results and inputs are considered as candidates, though as the
cache design evolves, the selected candidates may be refined. Ultimately we want
to cache and reuse as much as possible to minimize the computation costs. At the
same time, we want to limit the memory consumed by the cache.

### Initial Cache Candidates
Initial version caches only the right-hand side expressions of the
bindings. This way, we can guarantee that between the executions, the cache
stays within the memory limit without implementing [Memory
Management](#memorymanagement), the eviction strategy based on the overall size
of the cache.

### Further Development of Cache Candidates

> The next step is caching intermediate expressions. With this approach, we need a
> way to control the overall size of the cache because we can no longer guarantee
> the safe memory limits between the two program executions. Consider the example:

Reviewer: ```suggestion
The next step for the cache is to expand the portions of the introspected scope
that we cache. In general, this means the caching of intermediate expressions.

However, once we do this, we can no longer guarantee that we do not push the
JVM out of memory between two program executions. This is best demonstrated
by example. 
```

```
a = (computeHugeObject b).size
```

> Here we compute a considerably large object and discard it, saving only a single
> field - its size. Hence if we want to cache the intermediate result of discarded
> `computeHubeObject b` expression, we need to implement the mechanism of tracking
> the overall size of the cache. See [Memory Management](#memorymanagement)
> section.

Reviewer: ```suggestion
Here we compute a value that takes up a significant amount of memory, but from it we
only compute a small derived value (its size). Hence, if we want to cache the intermediate
result of the discarded `computeHugeObject b` expression, we need some way of tracking
the sizes of individual cache entries. 
```

2. If `k` is a dynamic symbol, all expressions that depend on _any instance_ of
   the dynamic symbol are evicted from the cache.

> Expressions that have been evicted from the cache subsequently have to be recomputed by the runtime.

Reviewer: This overflows.

Cache eviction takes into account the following aspects:

> Visualization. In the first place, we should care about nodes that have

Reviewer: This should be a bulleted list using the standard format with bolded headers.

visualization attached in the IDE. Priority. The runtime can assign a score to
a node, meaning how valuable this node is. Less valuable nodes should be evicted
first. Computation time. The runtime can calculate the time that node took to
compute. Less computationally intensive nodes should be evicted first. Memory

> limit. The cache should not exceed the specified memory limit.

Reviewer: This doesn't actually specify the eviction strategy. This document should contain a rigorous algorithmic spec of how we do eviction.

## Cache Backend

> The requirements for eviction strategies and memory management should guide the

Reviewer: This doesn't say anything useful.

exact implementation of the cache.

### Initial Implementation of Cache Backend

> The cache consists of key-value storage and an eviction function. Key-value

Reviewer: This is very light on detail. You need to describe the _what_ and the _why_ of the design and implementation of this cache. This section should also mention the details, such as the use of soft references as a fallback strategy.

storage can be as simple as a Hash Map with values wrapped into the Soft
References as a safety measure. Eviction function purges invalidated expressions
based on the [Cache Eviction Strategy](#cacheevictionstrategy).

### Further Development of Cache Backend

> The backend implementation should be updated according to the changing

Reviewer: First sentence here is useless. This section should instead be about how we plan on extending it in the future, not just examples. Furthermore it should have a standard `> The actionables for this section are` block, just as other documents do.

requirements of the runtime cache.For example, to extend the eviction strategy,
we can add an asynchronous scoring task that computes some properties of stored
objects (e.g., the size of the object). Those properties can be used in the
eviction strategy as optional clues, improving the hit ratio.

> ## Memory Management

Reviewer: This isn't a good title. It should be "Memory Bounded Caches" or similar.

Memory management refers to a way of controlling the size of the cache, avoiding
the Out Of Memory errors.

The methods below can be divided into two approaches.

> 1. Limiting overall JVM memory and relying on garbage collection

Reviewer: ```suggestion
1. Limiting the overall JVM memory and relying on garbage collection.
```

> 2. Calculating the object's size and using it in the eviction policy

Reviewer: ```suggestion
2. Calculating the object's size and using it in the eviction strategy.
```

In general, to control the memory size on JVM, anything besides limiting the

> total amount involves some tradeoff.

Reviewer: ```suggestion
total amount involves some tradeoffs.
```

### Soft References

> Soft References is a way to mark the cache entries available for garbage

Reviewer: This section should be broken up into multiple paragraphs.

collection whenever JVM runs a GC. In practice, it can cause long GC times and
reduced overall performance. This strategy is generally considered as a last
resort measure. The effect of the GC can be mitigated by using the _isolates_
(JSR121 implemented in GraalVM). One can think of an _isolate_ as a lightweight
JVM, running in a thread with their own heap, memory limit, and garbage
collection. The problem is that _isolates_ can't share objects. And even if we
move the cache to the separate _isolate_, that would require creating a
mechanism of sharing objects based on pointers, which requires implementing
serialization.  On the other hand, serialization itself can provide the size of
the object, which is enough to implement the eviction policy, even without
running the _isolates_.

### Serialization

> One approach to get the size of the value stored in the cache is

Reviewer: This section and its siblings are very light on detail. If you did benchmarking of these overheads please include numbers.

_serialization_. The downside is the computational overhead of transforming the
object into a byte sequence.

### Instrumentation
Another way of getting the object's size is to use JVM instrumentation. This
approach requires running JVM with _javaagent_ attached, which can complicate
the deployment and have indirect performance penalties.

### Manual Memory Management
This method implies tracking the size of the values by hand. It can be done for
the values of Enso language, knowing the size of its primitive types. For
interacting with other languages, Enso can provide an interface that developers
should implement to have a better experience with the cache.

> ### Comparison of Memory Management Approaches

Reviewer: This really should include actual numbers.
